src/java.base/share/classes/java/io/ObjectInputFilter.java

```diff
@@ -677,5 +1100,197 @@ public String toString() {
                 return pattern;
             }
         }
+
+        /**
+         * An ObjectInputFilter to evaluate a predicate mapping a class to a boolean.
+         */
+        private static class PredicateFilter implements ObjectInputFilter {
+            private final Predicate<Class<?>> predicate;
+            private final Status ifTrueStatus;
+            private final Status ifFalseStatus;
+
+            PredicateFilter(Predicate<Class<?>> predicate, Status ifTrueStatus, Status ifFalseStatus) {
+                this.predicate = predicate;
+                this.ifTrueStatus = ifTrueStatus;
+                this.ifFalseStatus = ifFalseStatus;
+            }
+
+            /**
+             * Apply the predicate to the class being deserialized, if the class is non-null
+             * and if it returns {@code true}, return the requested status. Otherwise, return UNDECIDED.
+             *
+             * @param info the FilterInfo
+             * @return the status of applying the predicate, otherwise {@code UNDECIDED}
+             */
+            public ObjectInputFilter.Status checkInput(FilterInfo info) {
+                return (info.serialClass() != null &&
+                        predicate.test(info.serialClass())) ? ifTrueStatus : ifFalseStatus;
+            }
+
+            public String toString() {
+                return "predicate(" + predicate + ")";
+            }
+        }
+
+        /**
+         * An ObjectInputFilter to evaluate if a FilterInfo is checking only limits,
+         * and not classes.
+         */
+        private static class AllowMaxLimitsFilter implements ObjectInputFilter {
```

Comment:
This class is maybe misnamed. If limitCheck == REJECTED it will not allow max limits. Or am I missing something?

Reply:
Rejection always wins in the larger scheme of things; another filter may reject based on other limits.
In the composition of filters, any UNDECIDED results must eventually be decided.
This filter maps, for a limit check, the UNDECIDED to allowed; it does nothing for checks for classes.
Other names considered,  allowUnlimited().  Also, not guaranteed.
Perhaps, something in the xxxElseYyy family.  Will reconsider the name.

Reply:
The allowMaxLimitsFilter will removed, its definition is confusing and it is not clear how it would be used.